### PR TITLE
 feat: Update README, tests, dependencies

### PR DESCRIPTION
**Pull request by germanattanasio**
@Naktibalda I did some updates to your code to update the tests to use typescript and cleanup extra code that we had for when we were using plain javascript.

* The code is now in `src` and `lib` is ignore from the repository.
* The npm package will have `lib` with the types and javascript but not `src` with the typescript.
* `simple-bot` was updated
* `multi-bot` was not updated as I ran out of time 😞
* switched to `jest` and added `codecov` for code coverage
* Removed dependencies that were not needed.

## What I need.
 * Can you please test that it works as expected for you?
 * I tried updating the README but I would like your 👀 there too.

---- for semantic versioning ---

BREAKING CHANGE:
* Botkit updated to 4.0.0 which means there are a lot of changes to use your bot with botkit now.
* Only simple bot example was updated.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -198,21 +223,19 @@ A common scenario of processing actions is:
...
 
-### using sendToWatson to update context (possible since v1.5.0)
+### Using sendToWatson to update context
 
 Using sendToWatson to update context simplifies the bot code compared to solution using updateContext below.
 
-```typescript
-function checkBalance(context, callback) {
+```js
+const checkBalance =  (context) => new Promise((resolve, reject) => {
```
Comment (Naktibalda) [inline README.md]: This function would look better with `async`.

File: README.md [hunk 2]
```diff
@@ -246,33 +268,24 @@ controller.on('message_received', processWatsonResponse);
 Events are messages having type different than `message`.
 
 [Example](https://github.com/howdyai/botkit/blob/master/examples/facebook_bot.js) of handler:
+
```
Comment (Naktibalda) [inline README.md]: Example link above is outdated and broken, here is the right one: https://github.com/howdyai/botkit/blob/master/packages/docs/reference/facebook.md#facebookeventtypemiddleware

File: README.md [hunk 3]
```diff
@@ -297,42 +310,42 @@ slackController.hears(['hello'], ['direct_message', 'direct_mention', 'mention']
...
 #### `before` and `after`
 
-The _before_ and _after_ callbacks can be used to perform some tasks _before_ and _after_ Assistant is called. One may use it to modify the request/response payloads, execute business logic like accessing a database or making calls to external services.
+The _before_ and _after_ async calls can be used to perform some tasks _before_ and _after_ Assistant is called. One may use it to modify the request/response payloads, execute business logic like accessing a database or making calls to external services.
 
 They can be customized as follows:
 
 ```js
-middleware.before = function(message, assistantPayload, callback) {
+middleware.before = (message, assistantPayload) => new Promise((resolve, reject) => {
```
Comment (Naktibalda) [inline README.md]:
```js
middleware.before = (message, assistantPayload) => async {
   // Code here gets executed before making the call to Assistant.
  return assistantPayload;
} 
```

File: examples/simple-bot/package.json [hunk 4]
```diff
@@ -8,9 +8,10 @@
   },
   "license": "Apache-2.0",
   "dependencies": {
-    "botkit": "^0.7.4",
-    "botkit-middleware-watson": "^1.8.4",
-    "dotenv": "^7.0.0",
+    "botbuilder-adapter-slack": "^1.0.1",
+    "botkit": "^4.0.1",
+    "botkit-middleware-watson": "file:../../botkit-middleware-watson-2.0.0.tgz",
```
Comment (Naktibalda) [inline examples/simple-bot/package.json]: Please change version constraint to `^2.0.0`.

**Comment by Naktibalda** [comment]
Is ignoring compiled files a standard practice in node-sdk?
It makes it impossible to install this library from branch because I get no `src/` and no `lib/`.

How am I supposed to test my changes now?
I would rather remove package-lock.json from git, because this is a library, so lock file doesn't matter.

File: src/index.ts [hunk 5]
```diff
@@ -15,13 +15,13 @@
...
 import AssistantV1 = require('ibm-watson/assistant/v1');
-import {Storage} from 'botbuilder';
-import {readContext, updateContext, postMessage, asCallback} from './utils';
+import { Storage } from 'botbuilder';
+import { readContext, updateContext, postMessage } from './utils';
 import deepMerge = require('deepmerge');
-import {promisify} from 'util';
+import { BotkitMessage } from 'botkit';
 
 export interface MiddlewareConfig {
```
Comment (Naktibalda) [inline src/index.ts]: I just used this in my code and I think that it would be better to name it `WatsonMiddlewareConfig`.

**Comment by germanattanasio** [comment]
@Naktibalda I added back the `lib` files so that you can test this from github.

File: README.md [hunk 6]
```diff
@@ -228,13 +228,13 @@ A common scenario of processing actions is:
...
 ```js
-const checkBalance =  (context) => new Promise((resolve, reject) => {
+const checkBalance =  async (context) => {
   //do something real here
   const contextDelta = {
     validAccount: true,
     accountBalance: 95.33
   };
-  resolve(context);
+  await Promise.resolve(context);
```
Comment (Naktibalda) [inline README.md]: This code can be further simplified to `return context` - async turns all returns to promises.

File: README.md [hunk 7]
```diff
@@ -326,16 +326,16 @@ The _before_ and _after_ async calls can be used to perform some tasks _before_
 They can be customized as follows:
 
 ```js
-middleware.before = (message, assistantPayload) => new Promise((resolve, reject) => {
-  // Code here gets executed before making the call to Assistant.
-  resolve(assistantPayload);
-});
+middleware.before = (message, assistantPayload) => async () => {
+   // Code here gets executed before making the call to Assistant.
+  return await Promise.resolve(assistantPayload);
```
Comment (Naktibalda) [inline README.md]: `return assistantPayload;`

File: README.md [hunk 8]
```diff
@@ -326,16 +326,16 @@ The _before_ and _after_ async calls can be used to perform some tasks _before_
...
+}
 ```
 
 ```js
-middleware.after = (message, assistantResponse) => new Promise((resolve, reject) => {
+middleware.after = (message, assistantResponse) => async () => {
   // Code here gets executed after the call to Assistant.
-  resolve(assistantResponse);
+  return await Promise.resolve(assistantResponse);
```
Comment (Naktibalda) [inline README.md]: `return assistantResponse;`

**Pull request merged**